### ISSUE TITLE
check for availability of `cudaMallocAsync`

### ISSUE THREAD

**Issue by janden** [flatironinstitute/finufft#445]
So it looks like this isn't always supported. In particular, it looks like the card in my laptop (GeForce 940MX, sm50) does not. This is fine, but should we be checking for this at some point and returning an error (see [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#stream-ordered-querying-memory-support)) or are we just relying on this trickling down to the user?

Unfortunately, I can't find a good rule of thumb outlining when this is supported (i.e. driver version, CUDA version, etc.). It seems that it varies from card to card so it makes it hard to specify our requirements beforehand.

Another solution would be to fall back to the malloc/free when this is not available, but I don't know what performance hit this would imply.

**Comment by ahbarnett** [comment]
@blackwer and @DiamonDinoia do you have any knowledge how to check for this? otherwise it crashes (seems to affect old GPUs).

**Comment by DiamonDinoia** [comment]
cudaMallocAsync is available from cuda 11.2 https://developer.nvidia.com/blog/enhancing-memory-allocation-with-new-cuda-11-2-features/

But to avoid the issue altogether could we just run the cudaMalloc in a background thread?

**Comment by janden** [comment]
> cudaMallocAsync is available from cuda 11.2 https://developer.nvidia.com/blog/enhancing-memory-allocation-with-new-cuda-11-2-features/

Right, but that's a sufficient. I'm running CUDA 12.4 locally and don't have it.

> But to avoid the issue altogether could we just run the cudaMalloc in a background thread?

If I understand correctly, this is what the stream framework is doing, but with the benefit that dependent operations (actually using the memory for something) is sequenced to after the memory is allocated.

**Comment by blackwer** [comment]
> If I understand correctly, this is what the stream framework is doing, but with the benefit that dependent operations (actually using the memory for something) is sequenced to after the memory is allocated.

This is correct. Using a background thread isn't quite the same. The `async` operations on the card allow for queuing up non-blocking operations on streams executing on the device, rather than doing that on the host. We should just probe the card for support and fall back on `cudaMalloc` if necessary.  `Kokkos` is having a similar issue right now: https://github.com/kokkos/kokkos/issues/7044 and they and NVIDIA suggest probing for it as noted in the linked comment. See https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html for the raw api.

I'll take a stab at this right now. @janden I doubt I have an affected card, so I might have to lean on you for testing.

**Comment by DiamonDinoia** [comment]
I see, basically there are two requirements:
1. compiler supports Async
2. device supports non blocking operations.

I like the solution in https://github.com/blackwer/finufft/commit/f7f41347cf6364dea146a4519a9168efd590cec0

Maybe we could still do the allocation in a background thread if Async is not supported? To overlap compute and transfers. This is the way I did it in the past when Async was not available. Is it a hassle to manage?

**Comment by blackwer** [comment]
> Maybe we could still do the allocation in a background thread if Async is not supported? To overlap compute and transfers. This is the way I did it in the past when Async was not available.

This should be possible, though quite a bit of a hassle. I think most unsupported devices are on their way out as obsolete anyway -- I don't think laptops shouldn't be our primary support concern, especially wrt to performance. I'm inclined to ignore it since the payoff is going to be pretty limited.

**Comment by DiamonDinoia** [comment]
Sounds good. The hassle does not outweigh the potential gains.

**Comment by DiamonDinoia** [comment]
Hi @blackwer,


For future reference, in case is useful. I remembered what I was doing in the past.
I was basically doing a cuda malloc and cuda memcpy together in a separate thread and I was relying on `--default-stream per-thread` for parallelism https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#default-stream

I did not need to have a queue or anything fancy to handle all of this as I was willing to accept the thread-creation destruction overhead.
